qubes/storage/lvm.py

```diff
@@ -33,19 +33,258 @@
+
+    return result
+
+
+def init_cache(log=logging.getLogger('qubes.storage.lvm')):
+    cmd = _init_cache_cmd
+    if os.getuid() != 0:
+        cmd = ['sudo'] + cmd
+    environ = os.environ.copy()
+    environ['LC_ALL'] = 'C.utf8'
+    p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+                         close_fds=True, env=environ)
+    out, err = p.communicate()
+    return_code = p.returncode
+    if return_code == 0 and err:
+        log.warning(err)
+    elif return_code != 0:
+        raise qubes.storage.StoragePoolException(err)
+
+    return _parse_lvm_cache(out)
+
+
+@asyncio.coroutine
+def init_cache_coro(log=logging.getLogger('qubes.storage.lvm')):
+    cmd = _init_cache_cmd
+    if os.getuid() != 0:
+        cmd = ['sudo'] + cmd
+    environ = os.environ.copy()
+    environ['LC_ALL'] = 'C.utf8'
+    p = yield from asyncio.create_subprocess_exec(*cmd,
+                                                  stdout=subprocess.PIPE,
+                                                  stderr=subprocess.PIPE,
+                                                  close_fds=True, env=environ)
+    out, err = yield from p.communicate()
+    return_code = p.returncode
+    if return_code == 0 and err:
+        log.warning(err)
+    elif return_code != 0:
+        raise qubes.storage.StoragePoolException(err)
+
+    return _parse_lvm_cache(out)
+
+
+size_cache_time = 0
+size_cache = init_cache()
+
+
+def _revision_sort_key(revision):
+    """Sort key for revisions. Sort them by time
+
+    :returns timestamp
+    """
+    if isinstance(revision, tuple):
+        revision = revision[0]
+    if '-' in revision:
+        revision = revision.split('-')[0]
+    return int(revision)
+
+
+def locked(method):
+    """Decorator running given Volume's coroutine under a lock.
+    Needs to be added after wrapping with @asyncio.coroutine, for example:
+
+    >>>@locked
+    >>>@asyncio.coroutine
+    >>>def start(self):
+    >>>    pass
+    """
+
+    @asyncio.coroutine
+    @functools.wraps(method)
+    def wrapper(self, *args, **kwargs):
+        with (yield from self._lock):  # pylint: disable=protected-access
+            return (yield from method(self, *args, **kwargs))
+
+    return wrapper
+
+
+def pool_exists(pool_id):
+    """ Return true if pool exists """
+    try:
+        vol_info = size_cache[pool_id]
+        return vol_info['attr'][0] == 't'
+    except KeyError:
+        return False
+
+
+def list_thin_pools():
+    """ Return list of thin pools """
+    thpools = []
+    try:
+        for key, vol in size_cache.items():
+            if vol['attr'][0] == 't':
+                # e.g. 'qubes_dom0/pool00'
+                parsed_key = key.split('/')
+                if len(parsed_key) == 2:
+                    volume_group = parsed_key[0]
+                    thin_pool = parsed_key[1]
+                    thpools.append((volume_group, thin_pool))
+    except KeyError:
```

Comment:
When KeyError can happen? Shouldn't it skip just that line then, instead of stopping processing everything?

Reply:
Yes it's ugly. I changed the check on `vol['attr']`. I don't remember if in some fake data I was not generating empty strings at some point in case of empty `vol['attr']`.